### ISSUE TITLE
Commentary about P6 packaging, and an invitation to discuss it further at package.community

### ISSUE THREAD

**Issue by raiph** [ugexe/zef#235]
(The following obviously isn't an issue as such. But this seemed like the right place to post this. Please accept my apologies if that was a mistake.)

Today there was [a twitter conversation](https://twitter.com/ReBeccaOrg/status/985402974364057600) (excerpted for your convenience below) about P6 packaging. Please consider commenting (here or on twitter or at [package.community](https://package.community/) or #perl6-toolchain or wherever works for you). TIA.

First and foremost Rebecca (they "work for npm on the cli" per [their twitter profile](https://twitter.com/ReBeccaOrg)) wrote:

> This kind of discussion, btw, is exactly why we created [package.community](https://package.community/)  (see the Discord in particular). I just haven't managed to lure any Perl folks over yet. 😉

Other than that, I've bundled most of the rest of what they wrote into four chunks that I think ought be helpful (but actually may or may not make sense and may or may not fairly represent what Rebecca was saying -- please navigate the twitter conversation itself if you wish to see the whole thing). Rebecca noted some specific things they found "disconcerting":

> Version data and author data at the per-module (in the perl6 sense) level. (Other langs modules are distributions in the perl sense.) ... Per-file is still too fine grained, I think, because… well I don't see how you'd correlate per-file w/ what you need to install? A ecosystem-wide index could be kept I guess?

> Global search path for modules, has potential to introduce dependency hell. ... what I saw that made my eyebrows raise was PERL6LIB and `use lib`, which amounts to the same thing, AFAICT [as "Global search path for modules" as @INC]

> While you can specify a version at use time, that creates a horrific maintenance burden if you ever need to change the version used across your app. ... While you might expect that bumping to say, a new major version would require changes everywhere anyway, in practice this is often not the case, particularly when adhering strictly to semver. ... having to select the version in source code wherever I require it that concerns me. ... I'm looking at this from the eyes of someone who's been in Node.js for years doing things I'd've historically done in Perl. The P6 model would push me toward consuming only a few large modules, the Nodejs model allows for extremely small modules. ... I try to imagine a small module ecosystem where you have to version every import in the code base (or risk creating dependency hell for your consumers)… and that's where my concerns over ongoing maintenance come in. ... To be clear: When I talk about maintenance burden I'm mostly talking about dev not ops. Both module author and module consumer though.

> I can't help shake feeling that dependency versions are better decoupled just a bit from code. ... the installer has to build a dependency graph built on distribution versions when the code is built on module versions and that interaction feels very wiggly to me. ... That multiple distributions can provide the same module (right?) and that doing so can have application spanning impacts and that also makes me very uncomfortable. (Where install order changes which module is used.)

**Comment by iarna** [comment]
👋

**Comment by raiph** [comment]
Discussion has continued on twitter. Selected further comments by Rebecca:

> I think META6.json is where dependency versions belong, not source code in modules, but there's no way to consume two modules w/ conflicting deps w/o having versions in the module source. ... And that means module authors need to ALWAYS include versions in their source that match their META6.json and that's what I find most awkward. This is a place where tooling could help, if the thing that built distribution tarballs rewrote the source to use versions everywhere… …then that would ease my concerns.

**Comment by ugexe** [comment]
Well, just because you *can* declare the version in your source code doesn't mean you should. The only reason to currently do this is if you want `myclass.^ver` to show something. Or more to the point: there is little reason to set the version inside of code for modules that one intends to package as a distribution.

`PERL6LIB` and `use lib` -- these are not really something to be used inside of a module source code one intends to distribute. Of course anyone is free to break those rules, but then one should not trust those modules. Additionally these two things are not quite the same, because in perl6 the order the dependencies are loaded matters due to precompilation. Indeed we give the user a machete for its utility and if they cut off a limb then they did something wrong then they should be more careful -- an alternative is we give the user a butter knife and see what kind of evil machinations they invent to fulfill these same purposes.

To summarize I think most of the concerns are over items that are not encouraged for module distributions, and are probably more "power user"-esq than anything ( meaning they should know what they are doing when using such things ).

**Comment by iarna** [comment]
@ugexe Ah, I think perhaps my commentary was not as clear as I intended, which is to say:

Let's just focus on one of my concerns to start with. I'm concerned that dependency hell (wanting to use two libs that require conflicting versions of the same module) is impossible to avoid without source-level versioning of every import, eg `use Foo` will always grab the latest version of Foo on the system, if two libs require two different versions they have to explicitly require the version they need to avoid getting the wrong version. They can't know in advance that a conflict like that exists (they're independent libs after all). As such, every module that's well behave HAS to declare a version when importing.

Worse becomes the default is that modules are machine global (and not per project), `use Foo` could import a newere, breaking release, installed by an unrelated project stored on the same system. Asking folks to set PERL6LIB _works_ but is, as I said at the start of my thread, not the kind of elegant solution I otherwise expect from p6.

In my perfect world, the module loader would read and understand the `META6.json` and use that as a default version when importing.

**Comment by ugexe** [comment]
This is all correct. Well behaved modules should be as explicit as possible ( `use Foo:ver<1>`, `use Foo:ver<1>:auth<github:bar>` ). For external dependencies I don't think this is asking too much.

I would agree it starts to become a problem when importing modules from their parent distribution. I've brought up the idea of automagically inferring these from `META6.json` `depends`. For instance lets consider a distribution `HTTP::UserAgent` which provides a `HTTP::Response` module and has an external dependency on `IO::Socket::SSL`:

```
unit module HTTP::UserAgent;

# external dependency
use IO::Socket::SSL:ver<...>:auth<...>:...<...>; # user *should* have to be explicit here

# provided module dependency
use HTTP::Response; # ideally this would expand to something like
                    # use HTTP::Response:ver(%?META<ver>):auth(%?META<auth>);
                    # instead of
                    # use HTTP::Response:ver<*>:auth<>;
```

Another alternative brought up was to mark such a same-distribution import differently:

```
use-my-stuff HTTP::Response;
```

Tooling could help, but would probably be a lot more work than implementing either suggestion above. So perl6 toolchain would like to find and implement a solution to this. As for why its not yet implemented: the subset of core perl6 developers and perl6 toolchain developers is currently just two people.

I hope this doesn't discourage you though, because I think you'll find the only reason we are lacking in certain areas here is because we have not had the tuits. We've discussed many a number of such NYI things in great details, but I'm afraid i'm not much of a record keeper outside of my own head.

**Comment by iarna** [comment]
At first glance `:auth<...>` was also making me a little twitchy (because singular authorship doesn't always match the dev model) but on second thought it's fine. It acts as a kind of name spacing. (If you can avoid a singular global pool of names that people gold-rush to, that's a huge advantage)

I also wonder about badly behaved distributions that contain modules claiming to be written by other authors? Are these just added wholesale to the machine global module repository?

> As for why its not yet implemented: the subset of core perl6 developers and perl6 toolchain developers is currently just two people.

Ha, yeah, I get that. The npm cli team is just me and one other person as well, and you all are responsible for a broader swath of tools. (Not that the Node.js tooling community is particularly larger. I'm not sure there _is_ an active tools team in core at the moment.)

> I hope this doesn't discourage you though, because I think you'll find the only reason we are lacking in certain areas here is because we have not had the tuits. We've discussed many a number of such NYI things in great details, but I'm afraid i'm not much of a record keeper outside of my own head.

I probably won't go jumping into P6 community participation _here_ as I'd want to wet my feet a bit more. (I've been thinking about starting with some tutorials for learning P6 from a modern JS point-of-view, in particular around asynchrony, promises and streams.)

**Comment by iarna** [comment]
Ah, and just occurring to me, but has any thought gone into how lock-files would work? How do you ensure that two systems will have exactly the same versions of modules and distributions? (Eg, from dev to CI to deployment…

It's something that we left to external tooling initially with npm, and then integrated but treated as external tooling for a long time and embraced in as a first-class feature the last year or so and we've seen a number of benefits.

**Comment by ugexe** [comment]
> I also wonder about badly behaved distributions that contain modules claiming to be written by other authors? Are these just added wholesale to the machine global module repository?

Basically it is up to the ecosystem to accept what it wants and the consumers to either trust that ecosystem or not. For instance a new p6c ecosystem could declare "only modules that list `github:username` as an auth that *also* corresponds to the github username of the person adding it to the ecosystem list will be accepted" -- users of zef could then configure to only query the ecosystems we consider trust-worthy. Since no truly trust-worthy ecosystem exists yet a bad actor could theoretically claim to be someone else, but at current scale this is still easily caught/remedied manually.

> Ah, and just occurring to me, but has any thought gone into how lock-files would work? How do you ensure that two systems will have exactly the same versions of modules and distributions? (Eg, from dev to CI to deployment…

At the moment I've not much thought about it. I suspect this will be one of the easier problems to solve once some of other stuff is implemented out better -- famous last words.

---

We also mulled over `META6.json` provides and depends acting as import selection filtering/ordering as mentioned earlier. The idea seems to be growing on us, but the scope of the changes required will need lots more thought and planning for how `CompUnit::Repository` should behave.

**Comment by niner** [comment]
The more we talk about managing large dependency trees and maintainability, the more I see 
@lizmat's wisdom when adding :api to our specification. This has a chance of being an even better tool than semver since it's absolutely explicit. You can't tell from a version 2.1.3 if the author follows semver or if it will work for a requested version 1.0+. But a module providing :api(2) will very probably not help when :api(1) is requested, i.e. a different api specifier is a clear signal that things were broken. :api is a better tool than versions because it explicitly has only a single, very specific job. We really should advertise its use as much as possible.

@iarna: This addresses the "wanting to use two libs that require conflicting versions of the same module" issue by rephrasing it: "wanting to use two libs that require conflicting api versions of the same module". That little change simplifies the issue a lot. Because `use Foo:api<1>` and `use Foo:api<2>` leave a lot of freedom to the module system for satisfying the dependency. For both api versions, we will load the highest installed version by default. And just in case not everyone knows: we can load and use both at the same time in Perl 6. So one can be explicit and avoid backwards compatibility issues while at the same time not being forced to pin down versions exactly.

Another critical but maybe underappreciated tool is Perl's strong culture of backwards compatibility. Our applications at work literally have hundreds of dependencies and yes, we really just install whatever comes from CPAN and there are practically no issues with that whatsoever. This is probably the reason why version pinning or lock files haven't been a priority so far for anyone. That's not to say that they don't have their place. It's just a less prominent one than in other languages.

@iarna: PERL6LIB or even more "use lib" are very useful tools, but each in their specific niche. And both are far away from code that's distributed. Except in tests. "use lib" frequently appears in test suits and it's absolutely fine there.

Of course, any kind of input (even more so if it contradicts whatever I write) from people who have dealt with large systems, dependency issues and package managers can only help us. So many thanks for that!

**Comment by raiph** [comment]
@iarna just tweeted:

> I _do_ think the thread can be closed. No one benefits from issues held open indefinitely.

So I'll close this issue with my own takeaway:

* Most of @iarna's concerns are over items that are not encouraged for module distributions.

* Most of the rest will be covered by implementing ideas already raised in this thread, especially using `META.json` and `:api`, when toolchainers have enough tuits.

**Comment by iarna** [comment]
@niner 

I like that `:api` gives a stronger hint as to what you're measuring than `:version`. It's interesting to be sure, though I'm not sure what it gets you over semver'd versions? It's fine though, not every language needs the same conventions.

> Another critical but maybe underappreciated tool is Perl's strong culture of backwards compatibility. 

I feel that culture of backwards compatibility isn't exactly a virtue but more mandated by the constraints of the Perl 5 module loader.

Back when I was a Perl  5 dev (basically for the 2000s and a little on either side), I rarely ran into backwards compatibility concerns, but I did hit them sometimes. Dependency hell (where you depend on dep A and dep B, and A and B both depend on different versions of C) was even more rare (and usually C libraries).

In Perl 5, IIRC, it was common practice to just release a brand new library when making backwards incompatible changes to the API. Or maybe not common practice, but certainly not uncommon. Major versions = breaking was a thing that happened sometimes, but was super painful because almost no one specified maximum versions for their dependencies, just minimum ones.

I also felt I was lacking a way of seeing what versions I _was_ using, that is, the full dependency graph was not very visible. Updating dependencies was … not a thing I generally _did_. I'd generally only get a new version of a lib if I installed something else that required the new version. Node.js was a huge cultural shift, where folks expect you to keep all of your dependencies upgraded all the time, and semver is the backstop against that flood of updates.

That being said, it has been rather refreshing to have the facility to make breaking changes and be ensured that your users won't accidentally update to them.

The problem with `:api<1>` is ... are you going to use that every time you ever import any module in your application? Are you going to do it for a module that only has one api version and doesn't currently appear likely to have more? What do you do when it updates? If I were doing P6 dev I'd want that as a best practice, but it's so much boilerplate that I have trouble imagining it actually becoming a cultural norm, particularly when not specifying it works just fine.

Regarding PERL6LIB and `use lib`, I agree, I even wrote [my own version](https://www.npmjs.com/package/@iarna/lib) of `use lib` for Node.js. I just don't believe they solve the concerns I have.

**Comment by raiph** [comment]
@iarna 

Thanks for your final(?) comment. Even if you don't reply to this I thought I'd respond to a couple of your points to see if @niner confirms I'm with the program.

If anyone prefers to reopen this, please do, but I think it's fine left closed.

> I'm not sure what [`:api`] gets you over semver'd versions? It's fine though, not every language needs the same conventions.

P6's design includes being radical polyglot. Think `use left-pad:from<JS>; use javax.xml.XMLConstants.XML_NS_PREFIX:from<Java>;`. To a degree, P6's packaging system must abstract over arbitrary ecosystems.

Hence an algebra over `:ver` *and* `:auth` *and* `:api` to (hopefully) conveniently and correctly cover semver *and* non-semver, correct *and* incorrect use of semver, backwards compatibility *and* moving on, official versions of a module *and* competing alternatives, and so on for a broad array of apparently mutually incompatible ecosystem designs, individual modules and packages, and realities on the ground.

Impossible? Well it's mostly just a vision at the moment, but I don't think it's one I've made up but rather one that emerged after I read the design documents and watched work and discussion by larry, lizmat, tobias, ugexe, niner and the rest of the gang over the last few years.

> I feel that culture of backwards compatibility isn't exactly a virtue but more mandated by the constraints of the Perl 5 module loader.

Spoken by a true JSer. :P

This is an interesting point. Computing is still young. Is caring about and supporting backwards compatibility relatively incompatible with caring about and supporting rapid language evolution?

Aiui the P6 perspective is that both are important and they can be reconciled to a degree and must be for P6 to do its thing. Maybe I'm wrong about that, or maybe I'm right but the P6 perspective needs to change?

> That being said, it has been rather refreshing to have the facility to make breaking changes and be ensured that your users won't accidentally update to them.

Imo this is the most important of your points.

(If P6 doesn't soon (10 years max) get to the point it can support rapid language evolution as well as backwards compatibility then it will have failed to deliver on what I considered one of it's most intriguing possibilities -- making it easier for its userland modules to drive both evolution and backwards compatibility among themselves and of P6's core.)

> The problem with `:api<1>` is ... are you going to use that every time you ever import any module in your application? ... it's so much boilerplate

I expect things like macros (one day, or should I say, one year), modules, and cultural and core language evolution to take care of any boilerplate issues.

Imo the issue for now is just whether we're choosing the right basic pegs to hang things on, as Larry might put it.

@niner, am I miles off?
